pipenv/environment.py

```diff
@@ -769,6 +769,8 @@ def is_satisfied(self, req: InstallRequirement):
             None,
         )
         if match is not None:
+            if req.link is None:
```

Comment:
Initial thoughts, I am wondering if this should actually return True since there was a match -- alternatively, we might handle this on line 780/782:  `elif req.link and match.has_metadata("direct_url.json")`

Reply:
Then we check on 772, 780 and 797 the boolean value of `req.link`, which could be done once before and the cut out from there. Maybe that is a style choice.

My idea was that if we can not properly compary e.g. that from 
```python
return (
    vcs_type == req.link.scheme
    and vcs_ref == requested_revision
    and direct_url_metadata["url"]
    == pipfile_part[req.link.scheme]
)
```

I would assume that expected behavour for not being able to make the comparison to `req.link.scheme`  should default to False.

But yes, thinking about it, it makes more sense to return True, as after the if statements True is also returned by default.